File: phi_plugin/model/cls/saveHistory.ts
import { Level } from "../constNum"
import LevelRecordInfo from "./LevelRecordInfo";
import Save from "./Save"
import { idString, levelKind } from "../type/type";
import { fCompute } from "..";

export default interface saveHistory {
    /**data货币变更记录 */
    data: { date: Date, value: number[] }[],
    /**rks变更记录 */
    rks: { date: Date, value: number }[],
    /**历史成绩 */
    scoreHistory: { [key: idString]: { [key in levelKind]?: [number, number, Date, boolean][] } },
    /**民间考核 */
    dan: any[],
    version: number
}

export default class saveHistory {

    /**
     * 
     * @param {saveHistory} data 
     */
    constructor(data: saveHistory) {

        for (let key in data?.data) {
            data.data[key].date = new Date(data.data[key].date)
        }
        this.data = data?.data || [];

        for (let key in data.rks) {
            data.rks[key].date = new Date(data.rks[key].date)
        }
        this.rks = data?.rks || [];

        this.scoreHistory = {}
        for (let key in data.scoreHistory) {
            let id = key.replace(/\.0$/, '')
            this.scoreHistory[id] = data.scoreHistory[key]
        }

        this.dan = data?.dan || [];
        /**v1.0,取消对当次更新内容的存储，取消对task的记录，更正scoreHistory */
        /**v1.1,更正scoreHistory */
        /**v1.2,由于曲名错误，删除所有记录，曲名使用id记录 */
        /**历史记录版本号 */
        this.version = data?.version

        /**检查版本 */
        if (!this.version || this.version < 2) {
            if (this.scoreHistory) {
                for (let i in this.scoreHistory) {
                    if (!i.includes('.0')) {
                        this.scoreHistory = {}
                    }
                    break
                }
            }
            this.version = 2
        }

    }

    /**
     * 合并记录
     * @param {saveHistory} data 另一个 History 存档
     */
    add(data: saveHistory) {
        this.data = merge(this.data, data.data)
        this.rks = merge(this.rks, data.rks)
        for (let song in data.scoreHistory) {
            if (!this.scoreHistory[song]) this.scoreHistory[song] = {}
            for (let dif in data.scoreHistory[song]) {
                if (this.scoreHistory[song] && this.scoreHistory[song][dif]) {
                    this.scoreHistory[song][dif] = this.scoreHistory[song][dif].concat(data.scoreHistory[song][dif])
                    this.scoreHistory[song][dif].sort((a, b) => {
                        return openHistory(a).date.getTime() - openHistory(b).date.getTime()
                    })
                } else {
                    this.scoreHistory[song][dif] = data.scoreHistory[song][dif]
                }
                let i = 1
                while (i < this.scoreHistory[song][dif].length) {
                    let last = openHistory(this.scoreHistory[song][dif][i - 1])
                    let now = openHistory(this.scoreHistory[song][dif][i])
                    if (last.score == now.score && last.acc == now.acc && last.fc == now.fc) {
                        // console.info(last.date.toISOString(), now.date.toISOString())
                        this.scoreHistory[song][dif].splice(i, 1)
                    } else {
                        ++i
                    }
                }
            }
        }
    }

    /**
     * 检查新存档中的变更并记录
     * @param {Save} save 新存档
     */
    update(save: Save) {
        // console.info(save)
        /**更新单曲成绩 */
        for (let id in save.gameRecord) {
            if (!this.scoreHistory[id]) this.scoreHistory[id] = {}
            for (let i in save.gameRecord[id]) {
                /**难度映射 */
                let level = Level[i]
                /**提取成绩 */
                let now: { score: number, acc: number, fc: boolean, date?: Date } = save.gameRecord[id][i]
                now.date = new Date(save.modifiedAt)
                /**本地无记录 */
                if (!this.scoreHistory[id][level] || !this.scoreHistory[id][level].length) {
                    this.scoreHistory[id][level] = [createHistory(now.acc, now.score, save.saveInfo.modifiedAt.iso, now.fc)];
                    continue
                }
                /**新存档该难度无成绩 */
                if (!save.gameRecord[id][i]) continue
                /**本地记录日期为递增 */
                for (let i = this.scoreHistory[id][level].length - 1; i >= 0; --i) {
                    /**第i项记录 */
                    let old = openHistory(this.scoreHistory[id][level][i])
                    // console.info(old.date.toISOString(), new Date(now.date).toISOString(), old.date.toISOString() == new Date(now.date).toISOString())
                    /**日期完全相同则认为已存储 */
                    if (old.score == now.score && old.acc == now.acc && old.fc == now.fc) {
                        /**标记已处理 */
                        now = null
                        break
                    }
                    /**找到第一个日期小于新成绩的日期 */
                    if (old.date < new Date(now.date)) {
                        /**历史记录acc仅保存4位，检查是否与第一个小于该日期的记录一致 */
                        if (old.acc.toString() != now.acc.toFixed(4) || old.score != now.score || old.fc != now.fc) {
                            /**不一致在第i项插入 */
                            this.scoreHistory[id][level].splice(i, 0, createHistory(now.acc, now.score, save.saveInfo.modifiedAt.iso, now.fc))
                        }
                        /**标记已处理 */
                        now = null
                        break
                    }
                }
                /**未被处理，有该难度记录，说明日期早于本地记录 */
                if (now) {
                    // console.info(11)
                    this.scoreHistory[id][level].unshift(createHistory(now.acc, now.score, save.saveInfo.modifiedAt.iso, now.fc))
                }
                /**查重 */
                let j = 1
                while (j < this.scoreHistory[id][level].length) {
                    let last = openHistory(this.scoreHistory[id][level][j - 1])
                    let now = openHistory(this.scoreHistory[id][level][j])
                    if (last.score == now.score && last.acc == now.acc && last.fc == now.fc) {
                        // console.info(last.date.toISOString(), now.date.toISOString())
                        this.scoreHistory[id][level].splice(j, 1)
                    } else {
                        ++j
                    }
                }
            }
        }
        /**更新rks记录 */
        for (let i = this.rks.length - 1; i >= 0; i--) {
            if (new Date(save.modifiedAt) > new Date(this.rks[i].date)) {
                if (!this.rks[i + 1] || (this.rks[i].value != save.saveInfo.summary.rankingScore || this.rks[i + 1]?.value != save.saveInfo.summary.rankingScore)) {
                    this.rks.splice(i + 1, 0, {
                        date: new Date(save.modifiedAt),
                        value: save.saveInfo.summary.rankingScore
                    })
                }
                break
            }
        }
        if (!this.rks.length) {
            this.rks.push({
                date: new Date(save.modifiedAt),
                value: save.saveInfo.summary.rankingScore
            })
        }
        /**更新data记录 */
        for (let i = this.data.length - 1; i >= 0; i--) {
            if (new Date(save.modifiedAt) > new Date(this.data[i].date)) {
                if (!this.data[i + 1] || (checkValue(this.data[i].value, save.gameProgress.money) && checkValue(this.data[i + 1]?.value, save.gameProgress.money))) {
                    this.data.splice(i + 1, 0, {
                        date: new Date(save.modifiedAt),
                        value: save.gameProgress.money
                    })
                }
                break
            }
        }
        if (!this.data.length) {
            this.data.push({
                date: new Date(save.modifiedAt),
                value: save.gameProgress.money
            })
        }

    }

    /**
     * 获取歌曲最新的历史记录
     * @param id 曲目id
     * @returns 
     */
    async getSongsLastRecord(id: idString) {
        let score = { ...this.scoreHistory[id] }
        let ans: { acc: number, score: number, date?: Date, fc: boolean }[] = []
        for (let i in score) {
            let level = Number(i)
            ans[level] = score[level] ? openHistory(score[level].at(-1)) : null
            if (!ans[level]) continue
            let date = ans[level].date
            ans[level] = new LevelRecordInfo(ans[level], id, level)
            ans[level].date = date
        }
        return ans
    }

    getRksAndDataLine() {

        let rks_history_ = []
        let data_history_ = []
        let user_rks_data = this.rks
        let user_data_data = this.data
        let rks_range = [17, 0]
        let data_range = [1e9, 0]
        let rks_date = [new Date(user_rks_data[0].date).getTime(), 0]
        let data_date = [new Date(user_data_data[0].date).getTime(), 0]

        for (let i in user_rks_data) {
            user_rks_data[i].date = new Date(user_rks_data[i].date)
            if (i as any <= 1 || user_rks_data[i].value != rks_history_[rks_history_.length - 2].value) {
                rks_history_.push(user_rks_data[i])
                rks_range[0] = Math.min(rks_range[0], user_rks_data[i].value)
                rks_range[1] = Math.max(rks_range[1], user_rks_data[i].value)
            } else {
                rks_history_[rks_history_.length - 1].date = user_rks_data[i].date
            }
            rks_date[1] = user_rks_data[i].date.getTime()
        }


        for (let i in user_data_data) {
            let value = user_data_data[i].value
            let valueNum = (((value[4] * 1024 + value[3]) * 1024 + value[2]) * 1024 + value[1]) * 1024 + value[0]
            user_data_data[i].date = new Date(user_data_data[i].date)
            if (i as any <= 1 || user_data_data[i].value != data_history_[data_history_.length - 2].value) {
                data_history_.push(user_data_data[i])
                data_range[0] = Math.min(data_range[0], valueNum)
                data_range[1] = Math.max(data_range[1], valueNum)
            } else {
                data_history_[data_history_.length - 1].date = user_data_data[i].date
            }
            data_date[1] = user_data_data[i].date.getTime()
        }

        let rks_history = []
        let data_history = []

        for (let i in rks_history_) {

            let ni = Number(i)

            if (!rks_history_[i + 1]) break
            let x1 = fCompute.percentage(rks_history_[ni].date, rks_date)
            let y1 = fCompute.percentage(rks_history_[ni].value, rks_range)
            let x2 = fCompute.percentage(rks_history_[ni + 1].date, rks_date)
            let y2 = fCompute.percentage(rks_history_[ni + 1].value, rks_range)
            rks_history.push([x1, y1, x2, y2])
        }

        for (let i in data_history_) {

            let ni = Number(i)

            if (!data_history_[ni + 1]) break
            let x1 = fCompute.percentage(data_history_[ni].date, data_date)
            let y1 = fCompute.percentage(data_history_[ni].value, data_range)
            let x2 = fCompute.percentage(data_history_[ni + 1].date, data_date)
            let y2 = fCompute.percentage(data_history_[ni + 1].value, data_range)
            data_history.push([x1, y1, x2, y2])
        }


        let unit = ["KiB", "MiB", "GiB", "TiB", "Pib"]

        for (let i in [1, 2, 3, 4]) {
            if (Math.floor(data_range[0] / (Math.pow(1024, Number(i)))) < 1024) {
                data_range[0] = `${Math.floor(data_range[0] / (Math.pow(1024, Number(i))))}${unit[i]}` as any
            }
        }

        for (let i in [1, 2, 3, 4]) {
            if (Math.floor(data_range[1] / (Math.pow(1024, Number(i)))) < 1024) {
                data_range[1] = `${Math.floor(data_range[1] / (Math.pow(1024, Number(i))))}${unit[i]}` as any
            }
        }

        return { rks_history, data_history, rks_range, data_range, rks_date, data_date }

    }

}

/**
 * 数组合并按照 date 排序并去重
 * @param m 
 * @param n 
 */
function merge(m: any[], n: any[]) {
    let t = m.concat(n)
    t.sort((a, b) => {
        return new Date(a.date).getTime() - new Date(b.date).getTime()
    })
    let i = 1
    while (i < t.length - 1) {
        /**因绘制折线图需要，需要保留同一值两端 */
        if (checkValue(t[i].value, t[i - 1].value) && checkValue(t[i].value, t[i + 1].value)) {
            t.splice(i, 1)
        } else {
            ++i
        }
    }
    return t
}
function createHistory(acc, score, date, fc) {
    return [acc.toFixed(4), score, date, fc]
}


/**
 * 展开信息
 * @param {Array} data 历史成绩
 */
function openHistory(data: any[]) {
    return {
        acc: Number(data[0]),
        score: Number(data[1]),
        date: new Date(data[2]),
        fc: Boolean(data[3])
    }
}

/**
 * 比较两个数组
 * @param {any} a 
 * @param {any} b 
 * @returns {boolean}
 */
function checkValue(a: any, b: any): boolean {
    /**非数组 */
    if (Object.prototype.toString.call(a) != '[object Array]') {
        return a == b
    }
    if (!a || !b) {
        return false
    }
    /**数组 */
    for (let i in a) {
        if (a[i] != b[i]) return false
    }
    return true
}

File: phi_plugin_js/model/cls/saveHistory.py
from datetime import datetime
import math
from typing import Any

from ...utils import Date
from ..constNum import MAX_DIFFICULTY, Level
from ..fCompute import fCompute
from .common import Save
from .LevelRecordInfo import LevelRecordInfo
from .models import HistoryModel, RecordModel, dataLine, rksLine, rksLineWithdataLine


def merge(m: list, n: list) -> list:
    """
    数组合并按照 date 排序并去重

    :param m: 第一个数组
    :param n: 第二个数组
    :return: 合并后的数组
    """
    t = m + n
    # 按照 date 排序
    t.sort(key=lambda x: x["date"])

    i = 1
    while i < len(t) - 1:
        # 因绘制折线图需要，需要保留同一值两端
        if checkValue(t[i]["value"], t[i - 1]["value"]) and checkValue(
            t[i]["value"], t[i + 1]["value"]
        ):
            t.pop(i)
        else:
            i += 1
    return t


def createHistory(
    acc: float, score: int, date: datetime, fc: bool
) -> tuple[float, int, datetime, bool]:
    """
    创建历史记录

    :param acc: 准确度
    :param score: 分数
    :param date: 日期
    :param fc: 是否全连
    :return: 历史记录列表
    """
    return (round(acc, 4), score, date, fc)


def openHistory(data: list | tuple) -> HistoryModel:
    """
    展开信息

    :param data:历史成绩
    """
    return HistoryModel(
        **{
            "acc": data[0],
            "score": data[1],
            "date": Date(data[2]),
            "fc": data[3],
        }
    )


def checkValue(a: Any, b: Any):
    """
    比较两个数组

    :param a: 第一个值
    :param b: 第二个值
    :return: bool
    """
    if not isinstance(a, list):
        return a == b

    if a is None or b is None:
        return False

    return all(a[i] == b[i] for i in range(len(a)))


class saveHistory:
    scoreHistory: dict[
        str,
        dict[
            str,
            list[tuple[float, int, datetime, bool]],
        ],
    ]
    """
    歌曲成绩记录
    ```
    {
        "songId": { - 曲目id
            "dif": [ - diff 难度 ["EZ", "HD", "IN", "AT", "LEGACY"]
                [acc:round(float, 4), score: int, date: datetime, fc: bool],
                [acc, score, date, fc],
                ...
            ]
        }
    }
    """
    data: list[RecordModel]
    """data货币变更记录"""
    rks: list[RecordModel]
    """rks变更记录"""
    challengeModeRank: list[RecordModel]
    """课题模式成绩"""
    version: float | None
    """
    历史记录版本号

    - v1.0,取消对当次更新内容的存储，取消对task的记录，更正scoreHistory
    - v1.1,更正scoreHistory
    - v2,由于曲名错误，删除所有记录，曲名使用id记录
    - v3,添加课题模式历史记录
    """
    dan: list
    """民间考核"""

    def __init__(self, data: dict[str, Any]):
        self.scoreHistory = data.get("scoreHistory") or {}
        self.data = data.get("data") or []
        self.rks = data.get("rks") or []
        self.challengeModeRank = data.get("challengeModeRank") or []
        self.version = data.get("version") or None
        self.dan = data.get("dan") or []

        # 检查版本
        if not self.version or self.version < 2:
            if self.scoreHistory:
                for i in self.scoreHistory:
                    if ".0" not in i:
                        self.scoreHistory = {}
                    break
            self.version = 2
        if self.version < 3:
            self.challengeModeRank = []
            self.version = 3

    def add(self, data: "saveHistory") -> "saveHistory":
        """
        合并记录

        :param saveHistory data: 另一个 History 存档
        """
        self.data = merge(self.data, data.data)
        self.rks = merge(self.rks, data.rks)
        self.challengeModeRank = merge(self.challengeModeRank, data.challengeModeRank)
        for song in data.scoreHistory:
            if not self.scoreHistory.get(song):
                self.scoreHistory[song] = {}
            for dif in data.scoreHistory.get(song, {}):
                if self.scoreHistory[song] and self.scoreHistory[song].get(dif):
                    self.scoreHistory[song][dif] = (
                        self.scoreHistory[song][dif] + data.scoreHistory[song][dif]
                    )
                    self.scoreHistory[song][dif].sort(key=lambda x: openHistory(x).date)
                else:
                    self.scoreHistory[song][dif] = data.scoreHistory[song][dif]
                i = 1
                while i < len(self.scoreHistory[song][dif]):
                    last = openHistory(self.scoreHistory[song][dif][i - 1])
                    now = openHistory(self.scoreHistory[song][dif][i])
                    if (
                        last.score == now.score
                        and last.acc == now.acc
                        and last.fc == now.fc
                    ):
                        self.scoreHistory[song][dif].pop(i)
                    else:
                        i += 1
        return self

    def update(self, save: Save):
        """
        检查新存档中的变更并记录

        :param Save save: 新存档
        """
        # 更新单曲成绩
        for id in save.gameRecord:
            if self.scoreHistory.get(id) is None:
                self.scoreHistory[id] = {}
                for i, _ in enumerate(save.gameRecord.get(id, [])):
                    # 难度映射
                    level = Level[i]
                    # 提取成绩
                    now = (
                        save.gameRecord[id][i] if i < len(save.gameRecord[id]) else None
                    )
                    if not now:
                        continue
                    now.date = save.saveInfo.modifiedAt.iso
                    # 本地无记录
                    if not self.scoreHistory[id].get(level):
                        self.scoreHistory[id][level] = [
                            createHistory(
                                now.acc,
                                now.score,
                                save.saveInfo.modifiedAt.iso,
                                now.fc,
                            )
                        ]
                        continue
                    # 新存档该难度无成绩
                    if i >= len(save.gameRecord[id]):
                        continue
                    # 本地记录日期为递增
                    for i in range(len(self.scoreHistory[id][level]) - 1, -1, -1):
                        # 第 i 项记录
                        old = openHistory(self.scoreHistory[id][level][i])
                        # 日期完全相同则认为已存储
                        if (
                            old.score == now.score
                            and old.acc == now.acc
                            and old.fc == now.fc
                        ):
                            # 标记已处理
                            now = None
                            break
                        # 找到第一个日期小于新成绩的日期
                        if old.date < now.date and (
                            old.acc != round(now.acc, 4)
                            or old.score != now.score
                            or old.fc != now.fc
                        ):
                            self.scoreHistory[id][level].insert(
                                i,
                                createHistory(
                                    now.acc,
                                    now.score,
                                    save.saveInfo.modifiedAt.iso,
                                    now.fc,
                                ),
                            )
                            # 标记已处理
                            now = None
                            break
                    # 未被处理，有该难度记录，说明日期早于本地记录
                    if now:
                        self.scoreHistory[id][level].insert(
                            0,
                            createHistory(
                                now.acc,
                                now.score,
                                save.saveInfo.modifiedAt.iso,
                                now.fc,
                            ),
                        )
                    # 查重
                    j = 1
                    while j < len(self.scoreHistory[id][level]):
                        last = openHistory(self.scoreHistory[id][level][j - 1])
                        now = openHistory(self.scoreHistory[id][level][j])
                        if (
                            last.score == now.score
                            and last.acc == now.acc
                            and last.fc == now.fc
                        ):
                            self.scoreHistory[id][level].pop(j)
                        else:
                            j += 1
        # 更新rks记录
        for i in range(len(self.rks) - 1, -1, -1):
            if save.saveInfo.modifiedAt.iso > self.rks[i].date:
                if (
                    i + 1 >= len(self.rks)
                    or self.rks[i].value != save.saveInfo.summary.rankingScore
                    or self.rks[i + 1].value != save.saveInfo.summary.rankingScore
                ):
                    self.rks.insert(
                        i + 1,
                        RecordModel(
                            **{
                                "date": save.saveInfo.modifiedAt.iso,
                                "value": save.saveInfo.summary.rankingScore,
                            }
                        ),
                    )
                break
        if not self.rks:
            self.rks.append(
                RecordModel(
                    **{
                        "date": save.saveInfo.modifiedAt.iso,
                        "value": save.saveInfo.summary.rankingScore,
                    }
                )
            )
        # 更新data记录
        for i in range(len(self.data) - 1, -1, -1):
            if save.saveInfo.modifiedAt.iso > self.data[i].date:
                if i + 1 >= len(self.data) or (
                    checkValue(self.data[i].value, save.gameProgress.money)
                    and checkValue(
                        self.data[i + 1].value if i + 1 < len(self.data) else None,
                        save.gameProgress.money,
                    )
                ):
                    self.data.insert(
                        i + 1,
                        RecordModel(
                            **{
                                "date": save.saveInfo.modifiedAt.iso,
                                "value": save.gameProgress.money,
                            }
                        ),
                    )
                break
        if not self.data:
            self.data.append(
                RecordModel(
                    **{
                        "date": save.saveInfo.modifiedAt.iso,
                        "value": save.gameProgress.money,
                    }
                )
            )
        # 更新课题模式记录
        for i in range(len(self.challengeModeRank) - 1, -1, -1):
            if save.saveInfo.modifiedAt.iso > self.challengeModeRank[i].date:
                clg = save.saveInfo.summary.challengeModeRank
                if clg != self.challengeModeRank[i].value and (
                    i + 1 > len(self.challengeModeRank)
                    or self.challengeModeRank[i + 1].value != clg
                ):
                    self.challengeModeRank.insert(
                        i + 1,
                        RecordModel(
                            **{
                                "date": save.saveInfo.modifiedAt.iso,
                                "value": save.saveInfo.summary.challengeModeRank,
                            }
                        ),
                    )
                break
        if not self.challengeModeRank:
            self.challengeModeRank.append(
                RecordModel(
                    **{
                        "date": save.saveInfo.modifiedAt.iso,
                        "value": save.saveInfo.summary.challengeModeRank,
                    }
                )
            )

    async def getSongsLastRecord(
        self, id: str
    ) -> dict[str, list[tuple[float, int, datetime, bool]]]:
        """
        获取歌曲最新的历史记录

        :param str id: 曲目id
        """
        t = {**self.scoreHistory.get(id, {})}
        for level in t:
            if records := t[level]:
                # 获取最新的一条记录并展开信息
                last_record = openHistory(records[-1])
                # 创建 LevelRecordInfo 实例
                level_info = await LevelRecordInfo.init(last_record, id, level)
                # 保留原始日期
                level_info.date = last_record.date
                t[level] = [level_info.to_tuple()]
            else:
                t[level] = []
        return t

    def getRksAndDataLine(self) -> rksLineWithdataLine:
        """折线图数据"""
        rks = self.getRksLine()
        data = self.getDataLine()
        return rksLineWithdataLine.from_models(rks, data)

    def getRksLine(self) -> rksLine:
        rks_range: list[float] = [MAX_DIFFICULTY, 0]
        rks_date: list[int] = []
        rks_history: list[list[float]] = []
        if user_rks_data := self.rks:
            rks_date = [int(user_rks_data[0].date.timestamp() * 1000), 0]
            rks_history_: list[RecordModel] = []
            for i, _ in enumerate(user_rks_data):
                user_rks_data[i].date = user_rks_data[i].date
                if (
                    i <= 1
                    or user_rks_data[i].value
                    != rks_history_[len(rks_history_) - 2].value
                ):
                    rks_history_.append(user_rks_data[i])
                    rks_range[0] = min(rks_range[0], user_rks_data[i].value)
                    rks_range[1] = max(rks_range[1], user_rks_data[i].value)
                else:
                    rks_history_[-1].date = user_rks_data[i].date
                rks_date[1] = int(user_rks_data[i].date.timestamp() * 1000)
            for i, _ in enumerate(rks_history_):
                if i + 1 >= len(rks_history_):
                    break
                x1 = fCompute.range(
                    int(rks_history_[i].date.timestamp() * 1000), rks_date
                )
                y1 = fCompute.range(rks_history_[i].value, rks_range)
                x2 = fCompute.range(
                    int(rks_history_[i + 1].date.timestamp() * 1000), rks_date
                )
                y2 = fCompute.range(rks_history_[i + 1].value, rks_range)
                rks_history.append([x1, y1, x2, y2])
            if not rks_history:
                rks_history.append([0, 50, 100, 50])
        return rksLine(
            **{
                "rks_history": rks_history,
                "rks_range": rks_range,
                "rks_date": rks_date,
            }
        )

    def getDataLine(self) -> dataLine:  # sourcery skip: extract-method
        data_range: list[float | str] = [1e9, 0]
        data_date: list[int] = []
        data_history: list[list[float]] = []
        if user_data_data := self.data:
            data_date = [int(user_data_data[0].date.timestamp() * 1000), 0]
            data_history_: list[RecordModel] = []
            for i, _ in enumerate(user_data_data):
                value = user_data_data[i].value
                assert isinstance(value, tuple)
                user_data_data[i].value = (
                    ((value[4] * 1024 + value[3]) * 1024 + value[2]) * 1024 + value[1]
                ) * 1024 + value[0]
                user_data_data[i].date = user_data_data[i].date
                if (
                    i <= 1
                    or len(data_history_) < 2
                    or user_data_data[i].value
                    != data_history_[len(data_history_) - 2].value
                ):
                    data_history_.append(user_data_data[i])
                    data_range[0] = min(data_range[0], user_data_data[i].value)
                    data_range[1] = max(data_range[1], user_data_data[i].value)
                else:
                    data_history_[len(user_data_data) - 1].date = user_data_data[i].date
                data_date[1] = int(user_data_data[i].date.timestamp() * 1000)
            for i, _ in enumerate(data_history_):
                if i + 1 >= len(data_history_):
                    break
                x1 = fCompute.range(data_history_[i].date.timestamp() * 1000, data_date)
                y1 = fCompute.range(data_history_[i].value, data_range)
                x2 = fCompute.range(
                    data_history_[i + 1].date.timestamp() * 1000, data_date
                )
                y2 = fCompute.range(data_history_[i + 1].value, data_range)
                data_history.append([x1, y1, x2, y2])
            unit = ["KiB", "MiB", "GiB", "TiB", "Pib"]
            for i in range(1, 5):
                assert isinstance(data_range[0], int)
                if math.floor(data_range[0] / (math.pow(1024, i))) < 1024:
                    data_range[0] = (
                        f"{math.floor(data_range[0] / (math.pow(1024, i)))}{unit[i]}"
                    )
            for i in range(1, 5):
                assert isinstance(data_range[1], int)
                if math.floor(data_range[1] / (math.pow(1024, i))) < 1024:
                    data_range[1] = (
                        f"{math.floor(data_range[1] / (math.pow(1024, i)))}{unit[i]}"
                    )
        return dataLine(
            **{
                "data_history": data_history,
                "data_range": data_range,
                "data_date": data_date,
            }
        )
